### ISSUE TITLE
Code style issues

### ISSUE THREAD

**Issue by GasperPaul** [sersajur/QPentago#15]
The code style is a mess.

- [ ] we have five different member functions naming conventions: 
    - Qt-style lower camel case (e.g., `GameState::getColCount`)
    - STLish lower snake case (`Game::new_game`)
    - upper snake case (`IView::New_game`)
    - upper camel case (`Game::GetPlayer`)
- [ ] we have underscores on private case names (`Referee::_AddCombination`) and double underscores explicitly forbidden by the Standard
- [ ] we have problem with file names (`GameState` is in `serialization.h` ffs!)
- [ ] minor indentation issues

@sersajur, @Youw can we discuss this?

**Comment by Youw** [comment]
Much to discuss?)) Just need to choose a single code style and stick to it.

A note from my side:
> minor indentation issues
I find mixed tabs and spaces - a major issue :)

**Comment by GasperPaul** [comment]
I mean, we must choose 😃 I'm for STL style or either camel case, in this order. And we probably should use linter on this.

I would consider major only something that would break the build. But yes, that is annoying and I want it fixed.

**Comment by Youw** [comment]
I prefer lower snake for variables and lower camel for functions.
I guess it worth discussing.

**Comment by GasperPaul** [comment]
I'm for it. And I think we had some logic with cases before, like do slots and signals be snake case or something, but it's baaaad.

**Comment by sersajur** [comment]
OK, lets discuss it.
I see code is really messy and for now any single convention is better than reality)
But if you are interested in my preferences:
- spaces over tabs [discussable]
- upper camel case for types [strongly recommended]
- lower camel case  or lower snake for variables [should select just one]
- lower camel case for functions [to be consistent with Qt here]
- lower snake case for namespaces [strongly recommended]
- m_ prefix for datamembers of classes [strongly recommended]
- i_, o_ prefixes for input and out parameters [strongly recommended]
- p_ prefix for pointers [discussable]
- use scope braces for single line code blocks under for, if, while, ..., statements [super strongly recommended]
- order of includes inside group lexicographical [strongly recommended]
- order of include groups: from more specific to less specific [discussable]

**Comment by Youw** [comment]
> spaces over tabs [discussable]

\+
> upper camel case for types [strongly recommended]

\+
> lower camel case or lower snake for variables [should select just one]

prefer lower snake case
> lower camel case for functions [to be consistent with Qt here]

\+
> lower snake case for namespaces [strongly recommended]

\+
> m_ prefix for datamembers of classes [strongly recommended]

against [discussable]
> i_, o_ prefixes for input and out parameters [strongly recommended]

strongly against:
1) input parameters by value or const reference (rarely by reference);
2) optional input parameters by pointer (to const);
3) output parameters by `return` statement in tuple or named struct;
4) optional output parameters - try to avoid, or by function overload;

If 1) - 4) &mdash; no prefixes ever required.

As an exception: if a large object is an output parameter and required to be cached outside a function call, it may be passed by reference, but, in most such cases, it is better to have an class instead of a function and have that cache-object as a member.

> p_ prefix for pointers [discussable]

strongly against
> use scope braces for single line code blocks under for, if, while, ..., statements [super strongly recommended]

agree for most cases, but I'd have a few exceptions for cases like:
```cpp
if (b == 0)
    throw std::runtime_error("Cannot divide by 0");
```
Generally, for `throw` and `return` statements.
> order of includes inside group lexicographical [strongly recommended]

\+
> order of include groups: from more specific to less specific [discussable]

I prefer groups (in order):
- OS headers;
- standard C headers (discussible, maybe merged with next group);
- standard C++ headers;
- 3d-party lib headers;
- our lib headers;
- current project headers (the only group with `#include "..."`)

In case if we choose  an already taken name by accident, the compiler will complain on our type/variable - not on some 3d-party one.

**Comment by sersajur** [comment]
> agree for most cases, but I'd have a few exceptions for cases like
well I don't like the idea to spend efforts to support these exceptions:
- replace content of `if` scope with `throw` or `return` - don't forget to remove braces;
- change `return` to assignment statement - don't forget to add braces;
- comment `return` or `throw` - don't forget comment `if` statement as well if you don't want to be surprised;
- need to add some logic before `return` under `if` - you have to add braces as well;
So lets be simple here, without exceptions, keeping uniformity.

**Comment by sersajur** [comment]
> In case if we choose an already taken name by accident, the compiler will complain on our type/variable - not on some 3d-party one.

Negative consequence of such manner is that if your header is lack of required standard library header - there is a high chance that its incompleteness will stay invisible.

**Comment by GasperPaul** [comment]
> * m_ - it is data member;

This is the most innocuous of the bunch, I'm ok with using this. There is an argument for `m_var` being better than `this->var` (at least it works better with `-Wshadow`, I believe?), though I still think that the number of cases where we really need to distinguish between local and member variables are too small.

> * o_, i_ - they are function arguments;

This is pure evil. Consider this:

```c++
void foo(const MyClass& i_ref) { }
void bar(MyClass& o_ref) {
    foo(o_ref); // are we lying about relation between o_ref and foo?
}```

Looking at function signature should be enough, and if you need to scroll too much then the function is just too long.

Also, in most cases it's easier to just return composite state objects/tuples. (Especially with structural bindings coming in C++17, woo-hoo!)

> * p_ prefix for pointers

I'm meh on this, but should it mark smart pointers too?

**Comment by sersajur** [comment]
> ```c++
> foo(o_ref); // are we lying about relation between o_ref and foo?
> ```
No, we don't lye here, since prefixes denote relations between _current_ function and its arguments.

But I'm OK if we don't use i_, o_, p_ at all.

**Comment by Youw** [comment]
> - replace content of if scope with throw or return - don't forget to remove braces;
> - change return to assignment statement - don't forget to add braces;
> - comment return or throw - don't forget comment if statement as well if you don't want to be surprised;
> - need to add some logic before return under if - you have to add braces as well;
> 
> So lets be simple here, without exceptions, keeping uniformity.

Never had any of those issues. Not that don't believe they are exist and real (I'm sure they are), but never hit any of those.
But I tend to agree with your suggestion. I guess it does less harm than it doesn't.

> Negative consequence of such manner is that if your header is lack of required standard library header - there is a high chance that its incompleteness will stay invisible.

Changing the order of includes does not guarantee making if visible (there might be other headers already included previously). It is just a half-solution.
I'd prefer rather always have a better compiler message, than maybe catch a missing include.

Speaking of last one: Qt Creator (starting with version 4.7, or something) has CLang frontend (and Static analyzer) enabled by-default as an engine for syntax highlighting and error checking. It literally always spots missing includes in headers, even if it doesn't prevent the whole project from being compiled.
I'm more convinced, that better use tools for self-checking (especially automated tools), rather than conventions.

> Qt Creator (...) has CLang frontend

I'm sure it is not the only IDE that supports it. Just choose by your preference.

**Comment by sersajur** [comment]
Deal! Let's organize includes to have better compiling error message. It will save our time, when CI build fail to compile.

In the same time, I believe, existence of built-in Qt Creator tools that solve some problems - this is not a reason to treat some problem as solved. I guess we should try organize things in IDE agnostic way.

**Comment by Youw** [comment]
There is a few more points, I'd like to raise:

- indentation: 4 spaces (except CMake - 2 spaces, as it is already consistent, historically I'd say);

- convention for constants? I prefer ALL_UPPER_CASE_SNAKE, but, for instance, [CppCoreGuidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-caps) does not recommend that with a pretty good reason;

Other points are easier on an example, than by words:
```cpp
//
// foo.h
class Foo
{ // brace on a new line
public:
    Foo();
    Foo(int val);

    void publicFunctions(int param1, int param);

//...

private:
    void privateFunc();

// suggestion - separate private functions and members
private:
    int private_member; // this one is not about the prefix
    double private_member2;
    std::string some_other_member { "prefer in-class initialization, if possible" };
};

//
// bar.h

//...

class Bar: public Foo // notice the spaces
{

};

//
// foo.cpp

//...

Foo::Foo()
{ } // no body at all

Foo::Foo(int param)
    : private_member(param) // linebreak and indentation
    , private_member2(param / 42)
{ // all open braces on a new line
    if (private_member < 0)
    {
        throw std::runtime_error("You have fucked this up!");
    }

}

```

**Comment by Youw** [comment]
- forward declare types in headers, to avoid unnecessary includes/dependencies in headers;

**Comment by sersajur** [comment]
> forward declare types in headers, to avoid unnecessary includes/dependencies in headers/speedup compile time;

sure. except types under std namespace, guess?

**Comment by sersajur** [comment]
> ALL_UPPER_CASE_SNAKE
I don't like that for constants. This case is already says "I am macro definition")

**Comment by Youw** [comment]
> sure. except types under std namespace, guess?

Maybe even more that that. Not much point forward declaring a small struct that lives in a header with no external dependencies.
I guess we need to follow our best judgments and don't get paranoia. 

> I don't like that for constants. This case is already says "I am macro definition")

That's why I'm opened to suggestions. `UpperCamelCase`?

**Comment by Youw** [comment]
> we should try organize things in IDE agnostic way

Agree. My point regarding Qt creator and CLang is quite the opposite:
It is not about "hey look, there is a new feature, we should try to use it because it is cool/everyone does it/etc.".
It is about "We have a problem, and this tool solves our problem. Lets use the tool (on a constant bases), because it makes us more productive/efficient/etc.".

**Comment by Youw** [comment]
Another point to raise:

- name semantics for classes/interfaces;
Suggest loose all prefixes, e.g. `I...`: ~"IView"~ -> "View";
Reason:
  - use **name semantic** for the same purpose. Meaning:
    Name `View` does not explain what kind on "view" it is. If it doesn't - it is an abstract "view". Having a name prefix e.g. "IView" is redundant and "noisy" - doesn't let us have clean code.
    Names like `OpenGLView` (or `ViewOpenGL`) or `ViewQGraphicItems` are self-explaining that they are concrete.

Naming like this will let the reader think of code he reads more in terms of "entities and their relations", and not get distracted by "interfaces and class hierarchies".

**Comment by GasperPaul** [comment]
> Other points are easier on an example, than by words

Let's expand that example and add somewhere into the docs.

> I don't like that for constants.

Same.

> That's why I'm opened to suggestions. `UpperCamelCase`?

I'm ok with that.

> Suggest loose all prefixes

I'm for this.
